### PR TITLE
ddl: fix wrong default value

### LINKED ISSUE #34881: Wrong default value for integer type
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Wrong default value for integer type cause error in tiflash: https://github.com/pingcap/tiflash/issues/3157
```mysql
mysql test> create table t (a int);
-- The type of column i is int, but the default value is set to "0.00"
mysql test> alter table t add column i int not null default 0.00; 
mysql test> show create table t;
```

### 2. What did you expect to see? (Required)
```mysql
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                          |
|       |   `a` int(11) DEFAULT NULL,                                 |
|       |   `i` int(11) NOT NULL DEFAULT '0'                       |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
```

### 3. What did you see instead (Required)
```mysql
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                          |
|       |   `a` int(11) DEFAULT NULL,                                 |
|       |   `i` int(11) NOT NULL DEFAULT '0.00'                       |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
```

### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
Signed-off-by: Lloyd-Pottiger <yan1579196623@gmail.com>

### What problem does this PR solve?

Issue Number: close #34881

Problem Summary:

### What is changed and how it works?
convert default value to standard format before set.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue that the default value might be wrong when sql is not normative.
```

### CODE REVIEW

**Review by xiongjiwei — COMMENTED** [review]
please also add some tests for it

**Comment by Lloyd-Pottiger** [comment]
> please also add some tests for it

done~

**Review by Defined2014 — APPROVED** [review]

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1160,6 +1160,11 @@ func getDefaultValue(ctx sessionctx.Context, col *table.Column, option *ast.Colu
 			// For BIT fields, convert int into BinaryLiteral.
 			return types.NewBinaryLiteralFromUint(v.GetUint64(), -1).ToString(), false, nil
 		}
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong:
+		// For integer fields, convert it into integer string literals. like convert "1.25" into "1" and "2.8" into "3".
+		if temp, err := v.ConvertTo(ctx.GetSessionVars().StmtCtx, &col.FieldType); err == nil {
+			v = temp
+		}
```
Comment (breezewish) [inline ddl/ddl_api.go:1167]: Maybe for other data types, like float, we should also do some conversion?

**Comment by Lloyd-Pottiger** [comment]
```mysql
mysql> create table t( a double NOT NULL DEFAULT 2.0 );
Query OK, 0 rows affected
Time: 0.079s

mysql> show create table t;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                               |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------+
| td    | CREATE TABLE `t` (\n  `a` double NOT NULL DEFAULT '2'\n) ENGINE=MyISAM DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci                  |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set
Time: 0.018s
```

**Comment by Lloyd-Pottiger** [comment]
Can anyone make a review and help merge this?

**Review by breezewish — APPROVED** [review]

**Pull request merged**